File: app/ClassScribe.Windows/tests/ClassScribe.Core.Tests/CaptureSignalHealthTests.cs
using ClassScribe.Core;

namespace ClassScribe.Core.Tests;

[TestClass]
public sealed class CaptureSignalHealthTests
{
    [TestMethod]
    public void NoCallbackBeforeBudgetStaysAwaiting()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        clock.NowSeconds = 4.99;

        Assert.AreEqual(CaptureSignalState.AwaitingCallbacks, tracker.Snapshot(attempt)?.State);
    }

    [TestMethod]
    public void NoCallbackAfterBudgetClassifiesUnavailable()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        clock.NowSeconds = 5;

        var snapshot = tracker.Snapshot(attempt);
        Assert.AreEqual(CaptureSignalState.NoCallbacks, snapshot?.State);
        Assert.AreEqual(0L, snapshot?.CallbackCount);
    }

    [TestMethod]
    public void ZeroSamplesAreSilentNotDead()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        Assert.IsTrue(tracker.TryRecordCallback(
            attempt,
            CaptureSignalMeasurement.FromRms(0, 0, 0)));

        var snapshot = tracker.Snapshot(attempt);
        Assert.AreEqual(CaptureSignalState.Silent, snapshot?.State);
        Assert.AreEqual(1L, snapshot?.CallbackCount);
        Assert.AreEqual(0L, snapshot?.SampleCount);
    }

    [TestMethod]
    public void SilentCallbacksKeepStreamHealthy()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        Assert.IsTrue(tracker.TryRecordCallback(attempt, SilentMeasurement));
        clock.NowSeconds = 2.99;
        Assert.IsTrue(tracker.TryRecordCallback(attempt, SilentMeasurement));

        var snapshot = tracker.Snapshot(attempt);
        Assert.AreEqual(CaptureSignalState.Silent, snapshot?.State);
        Assert.IsTrue(snapshot?.TransportIsHealthy == true);
    }

    [TestMethod]
    public void AudibleSamplesBecomeAudible()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        Assert.IsTrue(tracker.TryRecordCallback(attempt, AudibleMeasurement));

        var snapshot = tracker.Snapshot(attempt);
        Assert.AreEqual(CaptureSignalState.Audible, snapshot?.State);
        Assert.IsTrue(snapshot is not null && snapshot.EnergyDbfs > -40);
    }

    [TestMethod]
    public void AudibleToSilentIsNotFailure()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        Assert.IsTrue(tracker.TryRecordCallback(attempt, AudibleMeasurement));
        clock.NowSeconds = 1;
        Assert.IsTrue(tracker.TryRecordCallback(attempt, SilentMeasurement));

        var snapshot = tracker.Snapshot(attempt);
        Assert.AreEqual(CaptureSignalState.Silent, snapshot?.State);
        Assert.IsTrue(snapshot?.TransportIsHealthy == true);
    }

    [TestMethod]
    public void CallbackStallAfterHealthyStreamDetected()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);

        Assert.IsTrue(tracker.TryRecordCallback(attempt, AudibleMeasurement));
        clock.NowSeconds = 3;

        Assert.AreEqual(CaptureSignalState.NoCallbacks, tracker.Snapshot(attempt)?.State);
    }

    [TestMethod]
    public void StaleAttemptCannotChangeSignalState()
    {
        var clock = new FakeClock();
        var first = NewAttempt(1);
        var second = NewAttempt(2, first.SessionID);
        var tracker = NewTracker(clock, first);
        Assert.IsTrue(tracker.TryRecordCallback(first, AudibleMeasurement));

        clock.NowSeconds = 10;
        tracker.Begin(second);
        Assert.IsFalse(tracker.TryRecordCallback(first, SilentMeasurement));

        var snapshot = tracker.Snapshot(second);
        Assert.AreEqual(CaptureSignalState.AwaitingCallbacks, snapshot?.State);
        Assert.AreEqual(0L, snapshot?.CallbackCount);
        Assert.AreEqual(0L, snapshot?.SampleCount);
    }

    [TestMethod]
    public void NextAttemptStartsWithFreshHealthState()
    {
        var clock = new FakeClock();
        var first = NewAttempt(1);
        var second = NewAttempt(2, first.SessionID);
        var tracker = NewTracker(clock, first);
        Assert.IsTrue(tracker.TryRecordCallback(first, AudibleMeasurement));

        clock.NowSeconds = 10;
        tracker.Begin(second);

        var snapshot = tracker.Snapshot(second);
        Assert.AreEqual(CaptureSignalState.AwaitingCallbacks, snapshot?.State);
        Assert.AreEqual(0L, snapshot?.CallbackCount);
        Assert.AreEqual(0L, snapshot?.SampleCount);
        Assert.AreEqual(10d, snapshot?.StartedAtMonotonic);
    }

    [TestMethod]
    public void AsrEmptyDoesNotChangeCaptureHealth()
    {
        var clock = new FakeClock();
        var attempt = NewAttempt(1);
        var tracker = NewTracker(clock, attempt);
        Assert.IsTrue(tracker.TryRecordCallback(attempt, AudibleMeasurement));
        var before = tracker.Snapshot(attempt);

        var emptyTranscript = string.Empty;

        Assert.IsTrue(emptyTranscript.Length == 0);
        Assert.AreEqual(before, tracker.Snapshot(attempt));
    }

    private static CaptureSignalHealthTracker NewTracker(FakeClock clock, SessionAttemptID attempt)
    {
        var tracker = new CaptureSignalHealthTracker(CaptureSignalThresholds.Test, clock);
        tracker.Begin(attempt, clock.NowSeconds);
        return tracker;
    }

    private static SessionAttemptID NewAttempt(long generation, Guid? sessionID = null) =>
        SessionAttemptID.Create(sessionID ?? Guid.NewGuid(), generation);

    private static CaptureSignalMeasurement SilentMeasurement =>
        CaptureSignalMeasurement.FromRms(160, 160, 0);

    private static CaptureSignalMeasurement AudibleMeasurement =>
        CaptureSignalMeasurement.FromRms(160, 160, 0.2);

    private sealed class FakeClock : IMonotonicClock
    {
        public double NowSeconds { get; set; }
    }
}

File: app/MeetingTranscriber/ClassScribeTests/CaptureSignalHealthTests.swift
import AudioTapLib
import Foundation
import Testing
@testable import ClassScribe

private final class FakeCaptureMonotonicClock: @unchecked Sendable, CaptureMonotonicClock {
    var value: TimeInterval

    init(_ value: TimeInterval = 0) {
        self.value = value
    }

    func now() -> TimeInterval {
        value
    }

    func advance(by seconds: TimeInterval) {
        value += seconds
    }
}

private func silentMeasurement(samples: Int = 160) -> CaptureSignalMeasurement {
    CaptureSignalMeasurement(
        samples: [Float](repeating: 0, count: samples),
        channelCount: 1,
    )
}

private func audibleMeasurement(samples: Int = 160) -> CaptureSignalMeasurement {
    CaptureSignalMeasurement(
        samples: [Float](repeating: 0.2, count: samples),
        channelCount: 1,
    )
}

@Test
func noCallbackBeforeBudgetStaysAwaiting() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    clock.advance(by: 4.99)

    #expect(tracker.snapshot(for: attempt)?.state == .awaitingCallbacks)
}

@Test
func noCallbackAfterBudgetClassifiesUnavailable() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    clock.advance(by: 5)

    let snapshot = tracker.snapshot(for: attempt)
    #expect(snapshot?.state == .noCallbacks)
    #expect(snapshot?.callbackCount == 0)
}

@Test
func zeroSamplesAreSilentNotDead() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    #expect(tracker.recordCallback(
        for: attempt,
        measurement: CaptureSignalMeasurement(sampleCount: 0, frameCount: 0, rms: 0),
    ))

    let snapshot = tracker.snapshot(for: attempt)
    #expect(snapshot?.state == .silent)
    #expect(snapshot?.callbackCount == 1)
    #expect(snapshot?.sampleCount == 0)
}

@Test
func silentCallbacksKeepStreamHealthy() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    #expect(tracker.recordCallback(for: attempt, measurement: silentMeasurement()))
    clock.advance(by: 2.99)
    #expect(tracker.recordCallback(for: attempt, measurement: silentMeasurement()))

    let snapshot = tracker.snapshot(for: attempt)
    #expect(snapshot?.state == .silent)
    #expect(snapshot?.hasReceivedCallbacks == true)
    #expect(snapshot?.elapsedSinceLastCallback == 0)
}

@Test
func audibleSamplesBecomeAudible() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    #expect(tracker.recordCallback(for: attempt, measurement: audibleMeasurement()))

    let snapshot = tracker.snapshot(for: attempt)
    #expect(snapshot?.state == .audible)
    #expect(snapshot?.energyDBFS ?? -120 > -40)
}

@Test
func audibleToSilentIsNotFailure() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    #expect(tracker.recordCallback(for: attempt, measurement: audibleMeasurement()))
    clock.advance(by: 1)
    #expect(tracker.recordCallback(for: attempt, measurement: silentMeasurement()))

    let snapshot = tracker.snapshot(for: attempt)
    #expect(snapshot?.state == .silent)
    #expect(snapshot?.hasReceivedCallbacks == true)
}

@Test
func callbackStallAfterHealthyStreamDetected() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )

    #expect(tracker.recordCallback(for: attempt, measurement: audibleMeasurement()))
    clock.advance(by: 3)

    #expect(tracker.snapshot(for: attempt)?.state == .noCallbacks)
}

@Test
func staleAttemptCannotChangeSignalState() {
    let clock = FakeCaptureMonotonicClock()
    let first = SessionAttemptID(generation: 1)
    let second = SessionAttemptID(sessionID: first.sessionID, generation: 2)
    let tracker = CaptureSignalHealthTracker(
        attempt: first,
        thresholds: .test,
        clock: clock,
    )
    #expect(tracker.recordCallback(for: first, measurement: audibleMeasurement()))

    clock.advance(by: 10)
    tracker.begin(second)
    #expect(!tracker.recordCallback(for: first, measurement: silentMeasurement()))

    let snapshot = tracker.snapshot(for: second)
    #expect(snapshot?.state == .awaitingCallbacks)
    #expect(snapshot?.callbackCount == 0)
    #expect(snapshot?.sampleCount == 0)
}

@Test
func nextAttemptStartsWithFreshHealthState() {
    let clock = FakeCaptureMonotonicClock()
    let first = SessionAttemptID(generation: 1)
    let second = SessionAttemptID(sessionID: first.sessionID, generation: 2)
    let tracker = CaptureSignalHealthTracker(
        attempt: first,
        thresholds: .test,
        clock: clock,
    )
    #expect(tracker.recordCallback(for: first, measurement: audibleMeasurement()))

    clock.advance(by: 10)
    tracker.begin(second)

    let snapshot = tracker.snapshot(for: second)
    #expect(snapshot?.state == .awaitingCallbacks)
    #expect(snapshot?.callbackCount == 0)
    #expect(snapshot?.sampleCount == 0)
    #expect(snapshot?.startedAtMonotonic == 10)
}

@Test
func asrEmptyDoesNotChangeCaptureHealth() {
    let clock = FakeCaptureMonotonicClock()
    let attempt = SessionAttemptID(generation: 1)
    let tracker = CaptureSignalHealthTracker(
        attempt: attempt,
        thresholds: .test,
        clock: clock,
    )
    #expect(tracker.recordCallback(for: attempt, measurement: audibleMeasurement()))
    let before = tracker.snapshot(for: attempt)

    // An empty ASR result is deliberately not an input to this tracker. The
    // only accepted input is callback evidence from the capture adapter.
    let emptyTranscript = ""
    #expect(emptyTranscript.isEmpty)
    #expect(tracker.snapshot(for: attempt) == before)
}

@Test
func emptyLiveCallbackCountsAsTransportEvidence() async throws {
    let store = LiveAudioBufferStore()
    let generation = await store.reset()
    await store.append(
        LiveAudioBuffer(samples: [], channelCount: 2, sampleRate: 48_000, hostTime: 0),
        generation: generation,
    )

    #expect(try await store.waitForCallbacks(after: 0, timeout: 0) == 1)
    #expect(await store.totalSamples() == 0)
}
